runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineRunner.java

```diff
@@ -134,6 +138,69 @@ private PortablePipelineResult createPortablePipelineResult(
     }
   }
 
+  /**
+   * Patches all input coders of stateful ExecutableStage transforms. Stateful transforms aways have
+   * a KvCoder as input. Flink partitions the data based on the serialized version of the key. This
+   * key must retain the same binary representation to be able to serve the state in state requests
+   * from the SDK Harness from the correct partition. If the binary representation does not match,
+   * this will result in inconsistent checkpoints because keys are associated to a Flink key group
+   * which does not belong to the key group range of the state handling operator.
+   */
+  private static RunnerApi.Pipeline lengthPrefixKeyCoderOfStatefulStages(
```

Comment:
Can we move this logic into where we make an ExecutableStage in https://github.com/apache/beam/blob/b83b302ef97767e4ca245ea24e8bd40a6692e72c/runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/GreedyStageFuser.java#L65

All runners will should have have this.

Reply:
Should probably move this, unless other Runners do not care about the key encoding inconsistencies.

Strictly speaking, where it is now is the only correct place because if we apply combiner lifting from the Python SDK side, the fuser does not run: https://github.com/apache/beam/blob/119f41d2f8b1df1ad727d0ceac6f03a929cdb4b7/runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineRunner.java#L95

Reply:
It appears that this should be the "default" behavior to produce a consistent representation of the pipeline for runner consumption. Since the code here isn't Flink specific, it should move to a common package.

Reply:
`org.apache.beam.runners.core.construction.graph` ?

Reply:
I've moved this to `GreedyStageFuser`. Please have a look.

Moving it there has some other implications, i.e. that I also needed to patch Python to length prefix in its own optimizer/fuser.

Reply:
Actually, the patching of the Python SDK might not be necessary because combiners cannot contain user state / timers.